piv/jca.adoc

```diff
@@ -0,0 +1,247 @@
+----
+
+Inserting YKPiv provider in the first position makes it the preferred provider, and calling JCA APIs which don't specify a provider by name or pointer will return PIV JCA implementations (where applicable).
+
+The Provider uses services from existing other Providers to perform padding and hashing of messages.
+If additional Signature or Cipher schemes are required for use with YKPiv, you can install additional JCA Providers *prior* to instantiating the YKPiv PivProvider.
+For example, to be able to use an algorithm which is provided by the third party Provider Bouncy Castle, you could do the following:
+[source,java]
+
+----
+// This example installs an updated version of Bouncy Castle JCA provider
+// for this, we first need to remove the original Bouncy Castle provider
+Security.removeProvider("BC");
+// after the system Bouncy Castle provider has been removed, we install a new instance
+Security.addProvider(new BouncyCastle());
+
+// we call the PivProvider constructor after we have installed additional providers
+PivProvider pivProvider = new PivProvider(pivSession);
+Security.insertProviderAt(pivProvider, 1);
+----
+
+==== Alternative: Using the Provider directly
+
+To use YKPiv without global installation (avoiding call to `Security.insertProviderAt()`, the application needs to instantiate the provider class and use it in the JCA `getInstance` methods.
+For example:
+[source,java]
+
+----
+PivProvider pivProvider = new PivProvider(pivSession)
+KeyStore keyStore = KeyStore.getInstance("YKPiv", pivProvider);
+----
+
+=== Generating new keys
+
+The `KeyPairGenerator` service is used for generating new keys.
+Depending on requested key type, use the following code to acquire an instance of the service:
+
+[source,java]
+----
+// get instance of RSA KeyPairGenerator
+KeyPairGenerator rsaKpg = KeyPairGenerator.getInstance("YKPivRSA");
+
+// get instance of EC KeyPairGenerator
+KeyPairGenerator ecKpg = KeyPairGenerator.getInstance("YKPivEC");
+----
+
+The instance then needs to be initialized with `PivAlgorithmParameterSpec` object which describes the properties of the new key:
+[source,java]
+
+----
+PivAlgorithmParameterSpec(
+    Slot slot,
+    KeyType keyType,
+    @Nullable PinPolicy pinPolicy,
+    @Nullable TouchPolicy touchPolicy,
+    @Nullable char[] pin)
+----
+
+`slot`:: is the PIV slot, one of `Slot.AUTHENTICATION`, `Slot.SIGNATURE`, `Slot.KEY_MANAGEMENT` or `Slot.CARD_AUTH`
+`keyType`:: one of `KeyType.RSA1024`, `KeyType.RSA2048`, `KeyType.ECCP256`, `KeyType.ECCP384`; must match the `KeyPairGenerator` type
+`pinPolicy`:: defines the pin policy for using the key. `PinPolicy.DEFAULT`, `PinPolicy.NEVER`, `PinPolicy.ONCE` or `PinPolicy.ALWAYS`
+`touchPolicy`:: defines the whether or not a user presence is required (physical touch) when using the key.
+One of `TouchPolicy.DEFAULT`, `TouchPolicy.NEVER`, `TouchPolicy.ALWAYS`, `TouchPolicy.CACHED`
+`pin`:: the PIV application PIN
```

Comment:
I would mention here that the PIN only has to be supplied if the PrivateKey PinPolicy requires validation, and a private key operation (sign, decrypt) is going to be used on the resulting PrivateKey instance.